pkg/deploy/lattice/service_network_manager.go

```diff
@@ -43,6 +40,89 @@ type defaultServiceNetworkManager struct {
 	cloud pkg_aws.Cloud
 }
 
+func (m *defaultServiceNetworkManager) UpsertVpcAssociation(ctx context.Context, snName string, sgIds []*string) (string, error) {
+	sn, err := m.cloud.Lattice().FindServiceNetwork(ctx, snName, "")
+	if err != nil {
+		return "", err
+	}
+
+	associated, snva, _, err := m.isServiceNetworkAlreadyAssociatedWithVPC(ctx, *sn.SvcNetwork.Id)
+	if err != nil {
+		return "", err
+	}
+	if associated {
+		owned, err := m.cloud.CheckAndAcquireOwnership(ctx, *snva.Arn)
+		if err != nil {
+			return "", err
+		}
+		if !owned {
+			return "", services.NewConflictError("snva", snName,
+				fmt.Sprintf("Found existing vpc association not owned by controller: %s", *snva.Arn))
+		}
+		_, err = m.UpdateServiceNetworkVpcAssociation(ctx, &sn.SvcNetwork, sgIds, snva.Id)
+		if err != nil {
+			return "", err
+		}
+		return *snva.Arn, nil
+	} else {
+		req := vpclattice.CreateServiceNetworkVpcAssociationInput{
+			ServiceNetworkIdentifier: sn.SvcNetwork.Id,
+			VpcIdentifier:            &config.VpcID,
+			SecurityGroupIds:         sgIds,
+			Tags:                     m.cloud.DefaultTags(),
+		}
+		m.log.Debugf("Creating association between ServiceNetwork %s and VPC %s", sn.SvcNetwork.Id, config.VpcID)
+		resp, err := m.cloud.Lattice().CreateServiceNetworkVpcAssociationWithContext(ctx, &req)
+		if err != nil {
+			return "", err
+		}
+
+		serviceNetworkVpcAssociationStatus := aws.StringValue(resp.Status)
+		if serviceNetworkVpcAssociationStatus == vpclattice.ServiceNetworkVpcAssociationStatusActive {
+			return *resp.Arn, nil
+		} else {
+			m.log.Infof("Service network/vpc association is not in the active state. State is %s, will retry",
+				serviceNetworkVpcAssociationStatus)
+
+			return *resp.Arn, errors.New(LATTICE_RETRY)
+		}
+	}
+}
+
+func (m *defaultServiceNetworkManager) DeleteVpcAssociation(ctx context.Context, snName string) error {
+	sn, err := m.cloud.Lattice().FindServiceNetwork(ctx, snName, "")
+	if err != nil {
+		return err
+	}
+
+	associated, snva, _, err := m.isServiceNetworkAlreadyAssociatedWithVPC(ctx, *sn.SvcNetwork.Id)
```

Comment:
isServiceNetworkAlreadyAssociatedWithVPC deserve a change, 4 return values...
I believe we it can return (&SnVpcAssociation, error). nil means nothing found

Reply:
I'm too afraid of touching existing code, so I made a new (more concise IMO) function that does the job. I will clean this up once I release PR for default SN.